pyaerocom/colocation_3d.py

```diff
@@ -0,0 +1,492 @@
+    Some things you do not want to compute twice, however.
+    So (most of) the things that correspond to both colocation instances are computed here,
+    and then passed to the helper function.
+
+    Returns
+        colocated_data_lists : ColocatedDataLists
+
+    -------
+    """
+
+    if filter_name is None:
+        filter_name = const.DEFAULT_REG_FILTER
+    try:
+        data.check_dimcoords_tseries()
+    except DimensionOrderError:
+        data.reorder_dimensions_tseries()
+
+    var, var_aerocom = resolve_var_name(data)
+    if var_ref is None:
+        var_ref = var_aerocom
+        var_ref_aerocom = var_aerocom
+    else:
+        var_ref_aerocom = const.VARS[var_ref].var_name_aerocom
+
+    if not var_ref in data_ref.contains_vars:
+        raise VarNotAvailableError(
+            f"Variable {var_ref} is not available in ungridded "
+            f"data (which contains {data_ref.contains_vars})"
+        )
+    elif len(data_ref.contains_datasets) > 1:
+        raise AttributeError(
+            f"Colocation can only be performed with ungridded data objects "
+            f"that only contain a single dataset (input data contains: "
+            f"{data_ref.contains_datasets}. Use method `extract_dataset` of "
+            f"UngriddedData object to extract single datasets."
+        )
+
+    if any(
+        not {"start", "end"}.issubset(layer)
+        for layer in colocation_layer_limits + profile_layer_limits
+    ):
+        raise KeyError(
+            "start and end must be provided for profiles in each vertical layer in colocate_vertical_profile_gridded"
+        )
+
+    data_ref_meta_idxs_with_var_info = []
+    for i in range(len(data_ref.metadata)):
+        if not "altitude" in data_ref.metadata[i]["var_info"]:
+            logger.warning(
+                f"Warning: Station {data_ref.metadata[i]['station_name']} does not have any var_info"
+            )
+        else:
+            data_ref_meta_idxs_with_var_info.append(i)
+
+    if not all(
+        [
+            data.altitude.units == Unit(data_ref.metadata[i]["var_info"]["altitude"]["units"])
+            for i in data_ref_meta_idxs_with_var_info
+        ]
+    ):
+        raise DataUnitError
+        logger.info(
+            f"Mismatching units in colocation_3d.py. Model has units {data.altitude.units} whereas not all observations have this unit. Debug to find out where."
+        )
+
+    if update_baseyear_gridded is not None:
+        # update time dimension in gridded data
+        data.base_year = update_baseyear_gridded
+
+    # apply region filter to data
+    regfilter = Filter(name=filter_name)
+    data_ref = regfilter.apply(data_ref)
+    data = regfilter.apply(data)
+
+    # check time overlap and crop model data if needed
+    start, stop = check_time_ival(data, start, stop)
+    data = data.crop(time_range=(start, stop))
+
+    if regrid_res_deg is not None:
+        data = _regrid_gridded(data, regrid_scheme, regrid_res_deg)
```

Comment:
if these lines are nor covered by test, maybe the `if` clause should be marked with `# pragma: no cover`